src/notebooksyncagent.cpp

```diff
@@ -782,82 +824,193 @@ bool NotebookSyncAgent::applyRemoteChanges()
+        LOG_DEBUG("Ignore incidence already deleted locally:" << resource.href);
+        return false;
+    }
+
+    // find any existing incidence with this uid
+    QString incidenceIdentifier = QStringLiteral("%1 (%2)")
+            .arg(incidence->uid())
+            .arg(incidence->hasRecurrenceId() ? incidence->recurrenceId().toString() : QStringLiteral("original"));
+    mStorage->load(incidence->uid());
+    KCalCore::Incidence::Ptr storedIncidence;
+    switch (incidence->type()) {
+    case KCalCore::IncidenceBase::TypeEvent:
+        storedIncidence = mCalendar->event(incidence->uid(), incidence->hasRecurrenceId() ? incidence->recurrenceId() : KDateTime());
+        break;
+    case KCalCore::IncidenceBase::TypeTodo:
+        storedIncidence = mCalendar->todo(incidence->uid());
+        break;
+    case KCalCore::IncidenceBase::TypeJournal:
+        storedIncidence = mCalendar->journal(incidence->uid());
+        break;
+    case KCalCore::IncidenceBase::TypeFreeBusy:
+    case KCalCore::IncidenceBase::TypeUnknown:
+        qWarning() << "Unsupported incidence type:" << incidence->type();
+        return false;
+    }
+    if (storedIncidence) {
+        LOG_DEBUG("Updating existing event:" << incidenceIdentifier
+                                             << resource.href
+                                             << resource.etag);
+        storedIncidence->startUpdates();
+        IncidenceHandler::prepareImportedIncidence(incidence);
+        IncidenceHandler::copyIncidenceProperties(storedIncidence, incidence);
+        storedIncidence->setCustomProperty("buteo", "uri", resource.href);
+        storedIncidence->endUpdates();
+
+        // Save the modified incidence so it can be used to check whether there were
+        // local changes on the next sync.
+        mModifiedIncidenceICalData.insert(incidence->uid(), resource.iCalData);
+
+        // Save the incidence etag.
+        mUpdatedETags[incidence->uid()] = resource.etag;
+    } else {
+        LOG_DEBUG("Saving new event:" << incidenceIdentifier
+                                      << resource.href
+                                      << resource.etag);
+        KCalCore::Incidence::Ptr occurrence;
+        if (incidence->hasRecurrenceId()) {
+            // check to see if we had previously dissociated this occurrence from the recurring series.
+            occurrence = mCalendar->event(incidence->uid(), incidence->recurrenceId());
+            if (occurrence.isNull()) {
+ qWarning() << "no such recurring event occurrence?";
+                // no such dissociated occurrence existed, just create a new one.
+                // need to detach, and then copy the properties into the detached occurrence.
+                KCalCore::Incidence::Ptr recurringIncidence = mCalendar->event(incidence->uid());
+                occurrence = mCalendar->dissociateSingleOccurrence(recurringIncidence, incidence->recurrenceId(), incidence->recurrenceId().timeSpec());
```

Comment:
Appears ok to me. What the method does is clone the incidence without recurrence date, but with recurrence id  set, also an exception to main series rule is added.

On that last thing, I think the intention on standard is that recurrence-id has an implicit exception time on main series. Creating event on Outlook and inviting via email appears to generate .ics like that too. Then again Mkcal/Kcalcore doesn't appear to work like that. Not sure if it's ok or not to sync such exception back to remote server.